src/anml/parameter/parameter.py

```diff
@@ -0,0 +1,382 @@
+ParameterSets are sets of parameters that are related to each other, e.g. to parametrize the same
+distribution like mean and variance of the normal distribution. ParameterSets can also have
+functional priors.
+"""
+
+from dataclasses import fields, field
+from dataclasses import dataclass
+from typing import List, Callable, Optional, Dict
+import numpy as np
+import pandas as pd
+from copy import deepcopy
+
+from xspline import XSpline
+
+from anml.parameter.prior import Prior
+from anml.exceptions import ANMLError
+
+PROTECTED_NAMES = ['intercept']
+
+
+class VariableError(ANMLError):
+    pass
+
+
+class ParameterError(ANMLError):
+    pass
+
+
+class ParameterSetError(ANMLError):
+    pass
+
+
+@dataclass
+class Variable:
+    """A class that stores information about a variable.
+
+    Parameters
+    ----------
+    covariate: str
+        name of the covariate for this variable. 
+    var_link_fun: callable
+        link function for this variable. 
+    fe_init: float
+        initial value to be used in optimization for fixed effect. 
+    re_init: float
+        initial value to be used in optimization for random effect.
+    re_zero_sum_std: float, optional
+        standard deviation of zero sum prior for random effects.
+    fe_prior: Prior, optional
+        a prior of class :class:`~anml.parameter.prior.Prior`
+    re_prior: Prior, optional
+        a prior of class :class:`~anml.parameter.prior.Prior`
+
+    Attributes
+    ----------
+    All parameters become attributes after validation.
+    
+    """
+    var_link_fun: Callable = lambda x: x
+    covariate: str = None
+
+    fe_init: float = 0.
+    re_init: float = 0.
+
+    fe_prior: Prior = Prior()
+    re_prior: Prior = Prior()
+
+    re_zero_sum_std: float = field(default=np.inf)
+
+    def __post_init__(self):
+        if self.covariate in PROTECTED_NAMES:
+            raise VariableError("Choose a different covariate name that is"
+                                f"not in {PROTECTED_NAMES}.")
+
+    def design_mat(self, df: pd.DataFrame) -> np.ndarray:
+        """Returns the design matrix based on a covariate x.
+
+        Parameters
+        ----------
+        df
+            pandas DataFrame of covariate values (one dimensional)
+
+        Returns
+        -------
+        2-dimensional reshaped version of :python:`x`
+
+        """
+        x = df[self.covariate].values
+        return np.asarray(x).reshape((len(x), 1))
+
+
+@dataclass
+class Intercept(Variable):
+    """An intercept variable.
+    """
+    def __post_init__(self):
+        self.covariate = 'intercept'
+
+    def design_mat(self, df: pd.DataFrame) -> np.ndarray:
+        return np.ones((len(df), 1))
+
+
+@dataclass
+class Spline(Variable):
+    """A spline variable.
+    """
+
+    knots_type: str = 'frequency'
+    knots_num: int = 3
+    degree: int = 3
+    l_linear: bool = False
+    r_linear: bool = False
+
```

Comment:
would there be constraints for splines, like convexity, monotonicity, etc., be passed in in the future?

Reply:
good question -- @zhengp0 I think so right?

Reply:
Yes, but it is like the prior, depends on have do we want to incorporate the prior for `variable`, `parameter` and `parameters`